promising/logging_utils.py

```
@@ -0,0 +1,75 @@
+class PromisingHierarchyLogger:
+    def __init__(self, *, logger: logging.Logger | None = None, level: int):
+        self.logger = logger or logging.getLogger(type(self).__name__)
+        self.level = level
+
+    def log_awaiting_children(self, *, parent: "PromisingContext", children: Iterable["PromisingContext"]) -> None:
+        if not self.logger.isEnabledFor(self.level):
+            return
+
+        lines = ["AWAITING CHILDREN", self._fmt("parent", parent)]
+        for child in parent._active_children:
+            lines.append(self._fmt("direct child", child))
```

Ruff (0.15.9)

[warning] Use `list.extend` to create a transformed list

(PERF401)